### PR TITLE
Voice chat

### PR DESCRIPTION
**Pull request by bakscratch**
Adapted the base voice chat into razor with some styling.

image

### CODE REVIEW

File: code/UI/Hud.razor [hunk 1]
```diff
@@ -9,6 +9,9 @@
 	<HudStatus />
 	<Chatbox />
 	<TabMenu />
+	<TabMenu />
```
Comment (cr4yz) [inline code/UI/Hud.razor:12]: Duplicated TabMenu

File: code/UI/VoiceChat/StrafeVoiceChatEntry.razor [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+		Friend = new Friend(steamId);
+
+		SteamId(steamId);
+	}
+
+	public long SteamId(long id)
+	{
+		return id;
+	}
```
Comment (cr4yz) [inline code/UI/VoiceChat/StrafeVoiceChatEntry.razor:34]: What's this doing

File: code/UI/VoiceChat/StrafeVoiceChatEntry.razor [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+		}
+
+		Style.Dirty();
+	}
+
+	protected override int BuildHash()
+	{
+
+		return HashCode.Combine(Friend);
+	}
```
Comment (cr4yz) [inline code/UI/VoiceChat/StrafeVoiceChatEntry.razor:66]: Friend never changes after the constructor, don't think we need a BuildHash here.

File: code/UI/VoiceChat/StrafeVoiceChatEntry.razor [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+		var SpeakTimeout = 2.0f;
+		var timeoutInv = 1 - (timeSincePlayed / SpeakTimeout);
+		timeoutInv = MathF.Min(timeoutInv * 2.0f, 1.0f);
+
+		if (timeoutInv <= 0)
+		{
+			Delete();
+			return;
+		}
```
Comment (cr4yz) [inline code/UI/VoiceChat/StrafeVoiceChatEntry.razor:57]: Can this just be:
```
if(timeSincePlayed > 2f) 
{ 
    Delete(); 
    return; 
}
```

File: code/UI/VoiceChat/StrafeVoiceChatEntry.razor [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+		<label class="name" text="@(Friend.Name)"/>
+		<img class="avatar" src=@($"avatar:{Friend.Id}") />
+	</div>
+</root>
+
+@code
+{
+	public Friend Friend;
+
+	public long PlayerId;
```
Comment (cr4yz) [inline code/UI/VoiceChat/StrafeVoiceChatEntry.razor:15]: PlayerId not used anywhere, can it be removed?

File: code/UI/VoiceChat/StrafeVoiceChatEntry.razor [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+	private float VoiceLevel = 0.0f;
+	private float TargetVoiceLevel = 0;
+
+	RealTimeSince timeSincePlayed;
```
Comment (cr4yz) [inline code/UI/VoiceChat/StrafeVoiceChatEntry.razor:20]: timeSincePlayed casing and access modifier is inconsistent with the above

Comment (bakscratch) [reply code/UI/VoiceChat/StrafeVoiceChatEntry.razor:20]: this I don't fully understand, this is how it is now.
image

Comment (cr4yz) [reply code/UI/VoiceChat/StrafeVoiceChatEntry.razor]: This is just a minor aesthetic problem, there's two private variables there but typed out inconsistently.  TargetVoiceLevel is PascalCase, timeSincePlayed is camelCase, and one has private in front while the other doesn't:

```c#
private float TargetVoiceLevel;  // (note that float is 0 by default, no need to add = 0)
private RealTimeSince TimeSincePlayed;
```

```c#
float targetVoiceLevel;  // (note that float is 0 by default, no need to add = 0)
RealTimeSince timeSincePlayed;
```

I'm not too picky on whatever standards or conventions people like, but more just consistency.

Comment (bakscratch) [reply code/UI/VoiceChat/StrafeVoiceChatEntry.razor]: oh so they both should be the same basically?

Comment (bakscratch) [reply code/UI/VoiceChat/StrafeVoiceChatEntry.razor]: wait the TargetVoiceLevel can just be removed

**Pull request merged**